### PR TITLE
Adding Connection Timer

### PR DESCRIPTION
**Pull request by aruizExtraspace**
The current issue is the following:
The users tries to connect with a device, once the user presses the connect button the device was in range, but when the connect method is called in the DeviceManager in that point the device is unavailable. For this reason the code is not returning an answer.
This is a corner case but is important to have cover it to avoid unexpected behaviors when one app implements this SDK.

The solution was implement a timer in the connection method to know if the connection was success or not, if for some reason one of the delegates relates with the connection are not calling, the timer will trigger an error and report this error to the other layers.

Also is important to cancel the pending connection because once the device is available again the device will make the connection automatically this could cause some security issues because the user couldn't be in front of the lock

### CODE REVIEW

File: nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java [hunk 1]
```diff
@@ -1588,4 +1606,24 @@ public void useProxy(String address, int port){
...
+    private Runnable connectionTimerRunnable() {
+        return new Runnable() {
+            @Override
+            public void run() {
+                mGlobalNokeListener.onError(currentNoke, DEVICE_SHUTDOWN_RESULT, "Connection error");
+                disconnectNoke(currentNoke);
+            }
+        };
+    }
+
```
Comment (aruizExtraspace) [inline nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java:1628]: I'm open for suggestions on what kind of error you would send to the listener when the time finishes

File: nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java [hunk 2]
```diff
@@ -112,7 +114,18 @@
...
     private boolean mAllowAllDevices;
-
+    /**
+     *property used to detect if a connection fails when a device that is not available tries to create a connection
+     */
+    public Handler connectionTimer;
+    /**
+     *number of seconds the connection process will wait until return an error connection
+     */
+    public long numberOfSecondsToDetectTheConnectionError = 2;
```
Comment (aruizExtraspace) [inline nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java:124]: We are using 2 seconds here, but if you consider this should be different just let me know

File: nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java (hunk 1 above)
Comment (sapsley) [reply nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java:1628]: The `DEVICE_SHUTDOWN_RESULT` is a specific result packet from the lock firmware itself, so its usage should be reserved for that.

I recommend creating a new Noke Device Manager Service Error called `ERROR_CONNECTION_TIMEOUT = 317;` and using this.

File: nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java (hunk 2 above)
Comment (sapsley) [reply nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java:124]: Android devices often take longer to connect, so I recommend increasing this timeout. I'd go with a minimum of 6 secs. Maybe more if you start to see the timeout error popping up frequently during testing

File: nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java (hunk 1 above)
Comment (aruizExtraspace) [reply nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java:1628]: Thanks for the suggestion, making the change

File: nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java (hunk 2 above)
Comment (aruizExtraspace) [reply nokemobilelibrary/src/main/java/com/noke/nokemobilelibrary/NokeDeviceManagerService.java:124]: I'll create another PR with these changes, thanks

**Pull request merged**